### ISSUE TITLE
no_punct disables doc.sents

### ISSUE THREAD

**Issue by rxng** [chartbeat-labs/textacy#91]
Hi,

running text through the preprocess function with the no_punct filter makes doc.sents unusable. This is because it removes all the periods. However, the no_punct is necessary for things like -,!;: which compromise the analyses. Is there a way to get around this?

Thanks

**Comment by bdewilde** [comment]
Hi @rxng , unfortunately, there's no trivial way to resolve your issue. These pre-processing functions modify text _before_ you parse it, and there are warnings in the documentation about the problem you're facing ("be warned: preprocessing may affect the interpretation of the text -- and spacy's processing of it."). I tried not to limit users, since some people may only need extensive preprocessing of text for their own purposes, but I have a couple suggestions on how to get around this problem:

1. Do you _need_ to remove punctuation before parsing the text with spacy? If not, you can skip this particular preprocessing and do it post-spacy, e.g. `textacy.extract.words(doc, filter_punct=True)`.
1. I could add a "marks" param to the `remove_punct()` function, which would let you specify which punctuation marks, specifically, to remove. Then you could do something like `textacy.preprocess.remove_punct(text, marks='-,!;:')`. (Note that filtering exclamation points could also mess with spacy's sentence segmentation.)

Please let me know which solution works best for you. The second one would entail some work from me, but I'm happy to do it if that's a real user need.

**Comment by rxng** [comment]
Hi @bdewilde 

I must say I am shocked by your reply, in the best way possible - actually offering to add on things to meet user need! Can't say I have come across that much before!

I need to remove punctuation before putting the sentences into a list. Do you mean extracting as such?
`sentences = [textacy.extract.words(sents, filter_punct=True) for sents in doc.sent]`
Would it be resource intensive to call the function that many times?

I think the punctuation filtering could be very valuable. If we could choose what NOT to filter (i.e. period and exclamation marks), then we can retain the necessary format to later split into sentences.

R

**Comment by bdewilde** [comment]
Hi @rxng , I pushed [a commit](https://github.com/chartbeat-labs/textacy/commit/2fdabb14fd48534243b2e472995f4bec432ed493) that improves the remove_punct() function in a way that should resolve your problem. I opted to have users specify which punctuation they want to remove rather than which they _don't_ want, because that felt easier conceptually, but I'm not opposed to switching that around if you feel strongly about it.

Closing this issue, but don't hesitate to holler.
